catroid/src/main/java/org/catrobat/catroid/utils/CrashReporter.java

```diff
@@ -94,17 +88,9 @@ public static boolean logException(Throwable t) {
 	}
 
 	public static boolean checkIfCrashRecoveryAndFinishActivity(final Activity context) {
-		Log.d(TAG, "AFTER_EXCEPTION : checkIfCrashRecoveryAndFinishActivity()");
-		if (isRecoveredFromCrash()) {
-			if (isReportingEnabled()) {
-				sendUnhandledCaughtException();
-			}
-			if (!(context instanceof MainMenuActivity)) {
-				context.finish();
-			} else {
-				preferences.edit().putBoolean(RECOVERED_FROM_CRASH, false).commit();
-				return true;
-			}
+		if (isReportingEnabled()) {
+			sendUnhandledCaughtException();
+			return true;
 		}
 		return false;
 	}
```

Comment:
Not quite.

Everything related to RECOVERED_FROM_CRASH in checkIfCrashRecoveryAndFinishActivity ([ebce06f CrashReporter.java#L97](https://github.com/nishantt95/Catroid/blob/ebce06f4d6c4302c7a99c9edb71635190123dee5/catroid/src/main/java/org/catrobat/catroid/utils/CrashReporter.java#L97) and [ebce06f CrashReporter.java#L101-L106](https://github.com/nishantt95/Catroid/blob/ebce06f4d6c4302c7a99c9edb71635190123dee5/catroid/src/main/java/org/catrobat/catroid/utils/CrashReporter.java#L101-L106)) should be removed from the CrashReporter.java file and put somewhere else.

`CrashReporter.checkIfCrashRecoveryAndFinishActivity` is only called from `Utils.checkIfCrashRecoveryAndFinishActivity` which is only called from `BaseActivity.onCreate`. So to me it looks like moving the code related to RECOVERED_FROM_CRASH to the BaseActivity is the best place for two reason

* A method in Utils for that doesn't make sense since it's only used once and won't be used more often.
* The name checkIfCrashRecoveryAndFinishActivity indicates it does two things (note the *_And_* in the method name).

The remaining code in `CrashReporter.checkIfCrashRecoveryAndFinishActivity` is just the code from above. It mostly calls `sendUnhandledCaughtException`. However, there is still an if isReportingEnabled check which needs to be moved to sendUnhandledCaughtException as well.

And last, sendUnhandledCaughtException also needs to handle the case when EXCEPTION_FOR_REPORT is empty and `gson.fromJson(…)` may throw an exception.

So you need to update sendUnhandledCaughtException accordingly and write new tests.

Let me know if this clarifies it ^^